### ISSUE TITLE
Android task hijacking with taskAffinity,  Automatically add android:taskAffinity in a new project 

### ISSUE THREAD

**Issue by Wouter12** [flutter/flutter#63559]
## Use case

I discovered that Android task hijacking is allowed in a flutter application, any malicious app can put itself on the stack of the flutter application by adding `taskAffinity="com.example.myFlutterApp"` in the AndroidManifest.xml of the `<activity>` tag of the MainActivity

## Steps to Reproduce

2. create a new blank android project, in AndroidManifest.xml add 
`android:excludeFromRecents="true" android:launchMode="singleTask" android:taskAffinity="{com.example.myFlutterApp}"`  in the `<activity>` tag from the MainActivity.
1. open the `com.example.myflutterapp` application.
3. run the android app
4. close it and open the flutter application again

**Expected results:** 

The flutter application

**Actual results:** 
 
the android application

## Proposal

Automatically add `android:taskAffinity=""` to the `<application>` tag in the AndroidManifest.xml of the flutter app

**Comment by anjalsaneen** [comment]
Anti-virus software would detect as a malicious app if app id is not the same as task affinity.

**Comment by Hixie** [comment]
cc @reidbaker, see discussion in Discord

**Comment by stuartmorgan** [comment]
Isn't this https://developer.android.com/privacy-and-security/risks/strandhogg, which says the mitigation suggested here doesn't actually work, and lists changing the `compileSdkVersion` as the fix?

**Comment by gmackall** [comment]
From [discussion on discord](https://discord.com/channels/608014603317936148/610574672865656952/1211740382304538684) it seems like setting targetSdk may not actually mitigate this, and the only mitigation is actually not running on api levels <28 (AKA < Android 9). We are getting confirmation of this

**Comment by gmackall** [comment]
We've received confirmation that the second version of this vulnerability simply cannot be mitigated when running on devices using api levels below 28 (the [android docs](https://developer.android.com/privacy-and-security/risks/strandhogg) should be updated shortly). We've landed https://github.com/flutter/flutter/pull/144018, which should mitigate the first version of the attack. 

Other than dramatically shortening our support window for Android versions (https://docs.flutter.dev/reference/supported-platforms), there isn't anything more Flutter can do to help here, so I am going to close this issue.

And to clarify on the previous update after more discussion with John, 
> Our templates already specify targetSdkVersion >= 28 so any newly created Flutter app will be protected from this issue.

This is no longer true given the confirmation we received

> Gray is going to add some code to print a warning if flutter build sees a targetSdkVersion < 28.

I won't be adding a warning because the targetSdkVersion does not mitigate the attack

**Comment by Hixie** [comment]
I'm going to reopen this just for my own tracking.

I'm trying to understand exactly what the vulnerability is here.

What does the change we landed to the template do? If it's useful, should we not be warning developers who used our older template to make the change also?

Are there versions of Android where it is fundamentally impossible to secure an application from hijacking in this way? Should we be saying something on our supported platforms page about this?

**Comment by Hixie** [comment]
https://www.guardsquare.com/blog/protecting-against-strandhogg has some information that suggests there's more we can do.

**Comment by mariamhas** [comment]
> Are there versions of Android where it is fundamentally impossible to secure an application from hijacking in this way? Should we be saying something on our supported platforms page about this?

I think its a good idea to point people to the doc that the Android folks are updating about having a minsdk set being the only way to mitigate this issue in our support platforms page.  

Android doc: https://developer.android.com/privacy-and-security/risks/strandhogg

**Comment by johnmccutchan** [comment]
Yes, the existing official documentation has a bug in it that suggests setting "targetSdkVersion" is sufficient but it needs to be "minSdkVersion".

One way Flutter could protect its developers is to stop supporting Android < 29

**Comment by Hixie** [comment]
I think it's sufficient to warn people of the problem (not all applications will care about being spoofed).

It seems like we should detect apps that don't have the template's mitigation in place, and warn about the problem (is there any useful case where one would _not_ put the mitigation in place? If so we'd need a way to silence the warning, e.g. leaving a comment in the file that says it's intentionally not set).

We could consider what our default minimum should be, maybe with a comment in the file saying that we support down to a particular version but that certain versions have known vulnerabilities, pointing to the relevant documentation.

I agree with @mariamhas that documenting this on our web site's "supported platforms" page is good regardless.

**Comment by Hixie** [comment]
- [ ] Warn people when their `AndroidManifest.xml` file doesn't contain the string `android:taskAffinity`.
- [ ] Consider if our template minimum should be above our supported minimum, with a comment explaining why.
- [ ] Document that older Android versions have unmitigatable security risks on our supported platforms page.

**Comment by Hixie** [comment]
I'm going to close this issue in favor of the ones filed above.
